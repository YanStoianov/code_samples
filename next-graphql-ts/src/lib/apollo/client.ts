import { ApolloClient, InMemoryCache, HttpLink, ApolloLink, concat } from "@apollo/client";
import { CookieValueTypes, getCookie } from "cookies-next";
import { userConstants } from "@/store/user/constants";
import cacheOptions from "./cacheOptions";

interface ApolloClientProps {
  token?: CookieValueTypes;
}

const httpLink = new HttpLink({ uri: process.env.NEXT_PUBLIC_APOLLO_URI, includeExtensions: true });
const authToken = getCookie(userConstants.ACCESS_TOKEN);

const authMiddleware = ({ token }: ApolloClientProps) => {
  return new ApolloLink((operation, forward) => {
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    operation.setContext(({ headers }: Record<string, any>) => {
      return {
        headers: {
          authorization: token ? `Bearer ${token}` : "Bearer guest",
          ...headers,
        },
      };
    });
    return forward(operation).map((response) => {
      if (response.data) {
        response.data.extensions = response.extensions;
      }
      return response;
    });
  });
};

const client = new ApolloClient({
  ssrMode: true,
  link: concat(authMiddleware({ token: authToken }), httpLink),
  cache: new InMemoryCache(cacheOptions),
});

export const getApolloClient = ({ token }: ApolloClientProps) => {
  return new ApolloClient({
    ssrMode: true,
    link: concat(authMiddleware({ token }), httpLink),
    cache: new InMemoryCache(cacheOptions),
  });
};

export default client;
