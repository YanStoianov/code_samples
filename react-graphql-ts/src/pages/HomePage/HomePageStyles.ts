import styled from "styled-components";

export default styled.div`
  padding: 10px 30px;

  .pika-table {
    th {
      text-transform: capitalize;
    }

    tr td:first-child {
      text-transform: capitalize;
    }
  }

  .name-block {
    display: flex;
    justify-content: center;
    align-items: center;
    flex-direction: column;

    img {
      width: 100%;
    }
  }
`;