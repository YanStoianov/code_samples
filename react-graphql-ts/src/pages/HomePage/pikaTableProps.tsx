import {ColumnsType} from "antd/es/table";
import React from "react";
import {PokemonSpecy, PokemonStat, PokemonTypeGroup} from "../../interfaces/pokeInterfaces";
import {titleCase} from "../../utils/stringUtils";
import {CheckCircleOutlined, CloseCircleOutlined} from "@ant-design/icons";

export const generateStatColumns = (stats: string[]) => {
    return stats.map(stat => (
        {
            title: stat,
            dataIndex: 'pokemon_v2_pokemonstats',
            key: `pokemon_v2_pokemontypes${stat}`,
            width: '120px',
            sorter: (a: any, b: any) => {
                const statA = a.pokemon_v2_pokemonstats
                    .find((statItem: PokemonStat) => statItem.pokemon_v2_stat.name === stat)?.base_stat || 0
                const statB = b.pokemon_v2_pokemonstats
                    .find((statItem: PokemonStat) => statItem.pokemon_v2_stat.name === stat)?.base_stat || 0

                return statA - statB;
            },
            render: (statList: PokemonStat[]) => {
                const resultStat = statList.find((statItem) => statItem.pokemon_v2_stat.name === stat)
                return resultStat?.base_stat
            }
        }
    ))
}

interface DataType {
    key: React.Key;
    name: string;
    id: number;
    pokemon_v2_pokemontypes: [];
}


export const columns: ColumnsType<DataType> = [
    {
        title: 'Name',
        dataIndex: 'name',
        defaultSortOrder: 'ascend',
        width: '200px',
        sorter: (a, b) => a.name.localeCompare(b.name),
        render: (name, {id}) => {
            return (
                <div className={'name-block'}>
                    <img
                        src={`https://raw.githubusercontent.com/PokeAPI/sprites/master/sprites/pokemon/${id}.png`}
                        alt={name}
                    />
                    <div>{name}</div>
                </div>
            );
        }
    },
    {
        title: 'Types',
        dataIndex: 'pokemon_v2_pokemontypes',
        width: '100px',
        render: (typeList) => {
            const types = typeList.map((item: PokemonTypeGroup) => {
                return item['pokemon_v2_type']['name'];
            }).join(',');
            return <div>{types}</div>;
        }
    },
    ...generateStatColumns(['hp', 'attack', 'defense', 'special-attack', 'special-defense', 'speed']),
    {
        title: 'Species',
        dataIndex: 'pokemon_v2_pokemonspecy',
        width: '450px',
        render: (species: PokemonSpecy) => {
            const speciesList: PokemonSpecy = {
                ...species,
                generation: species.pokemon_v2_generation?.name,
                growthrate: species.pokemon_v2_growthrate?.name
            };
            delete speciesList.pokemon_v2_growthrate;
            delete speciesList.pokemon_v2_generation;

            return <table className={'species-table'}>
                <tbody>
                {
                    Object.keys(speciesList).filter(key => key !== '__typename').map((key) => {
                        const value = (speciesList as any)[key];
                        let specyValue: JSX.Element = <>{JSON.stringify(value)}</>

                        if (value === true) {
                            specyValue = <CheckCircleOutlined className={'approved'}/>
                        }
                        if (value === false) {
                            specyValue = <CloseCircleOutlined className={'disapproved'}/>
                        }


                        return (
                            <tr key={key}>
                                <td className={'specy-name'}><b>{titleCase(key)}</b></td>
                                <td className={'specy-value'}>{specyValue}</td>
                            </tr>
                        )
                    })
                }
                </tbody>
            </table>
        }
    }
];