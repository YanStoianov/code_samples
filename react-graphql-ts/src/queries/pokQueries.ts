import {gql} from "@apollo/client";

export const GET_POKEMONS = gql`
    query GetPokemons($nameSearch: String!) {
        pokemon_v2_pokemon(where: {name: {_ilike: $nameSearch}}) {
            id, name,
            pokemon_v2_pokemontypes {
                pokemon_v2_type {
                    id,
                    name
                }
            },
            pokemon_v2_pokemonstats {
                id
                pokemon_v2_stat {
                    name
                }
                base_stat
            },
            pokemon_v2_pokemonspecy {
                base_happiness
                capture_rate
                forms_switchable
                gender_rate
                has_gender_differences
                hatch_counter
                is_baby
                is_legendary
                is_mythical
                pokemon_v2_generation {
                    name
                    id
                }
                pokemon_v2_growthrate {
                    name
                    id
                }
            }
        }
    }
`;

export const GET_POKEMONS_COUNT = gql`
    query GetPokemonsCount($default: Boolean!) {
        pokemon_v2_pokemon_aggregate(where: {is_default: {_eq: $default}}) {
            aggregate {
                count
            }
        }
    }
`;

export const GET_POKEMON_TYPES = gql`
    query GetPokemonTypes {
        pokemon_v2_type {
            id, name
        }
    }
`;
