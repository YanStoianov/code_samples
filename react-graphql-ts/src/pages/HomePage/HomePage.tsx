import React, {useMemo, useState} from 'react';
import MainLayout from '../../layouts/MainLayout';
import {useQuery} from "@apollo/client";
import {GET_POKEMON_TYPES, GET_POKEMONS} from "../../queries/pokQueries";
import {Table} from 'antd';
import LoaderWrapper from "../../components/LoaderWrapper/LoaderWrapper";
import HomePageStyles from "./HomePageStyles";
import {columns} from "./pikaTableProps";
import {
    AdditionalPokeFilters, boolPokeFilters,
    minMaxPokeFilters,
    PokemonInterface,
    PokemonTypes
} from "../../interfaces/pokeInterfaces";
import PokeStats from "../../components/PokeStats/PokeStats";
import PokeFilter from "../../components/PokeFilter/PokeFilter";
import inRange from "lodash/inRange";

const minMax = {
    min: 0,
    max: 0
}
export const initialAdditionalFilters: AdditionalPokeFilters = {
    base_happiness: {...minMax},
    capture_rate: {...minMax},
    gender_rate: {...minMax},
    hatch_counter: {...minMax},
    has_gender_differences: false,
    forms_switchable: false,
    is_legendary: false,
    is_mythical: false,
    generation: '',
    growthrate: '',
    is_baby: false,
    useAdditionalFilters: false
}

const HomePage: React.FC = () => {
    const [nameSearch, setNameSearch] = useState('');
    const [selectedTypes, setSelectedTypes] = useState<number[]>([]);
    const [additionalFilters, setAdditionalFilters] = useState<AdditionalPokeFilters>(initialAdditionalFilters);

    const {loading, data: pokemonData} = useQuery(GET_POKEMONS, {
        variables: {
            nameSearch: `%${nameSearch}%`
        }
    });
    const {data: pokemonTypes} = useQuery<PokemonTypes>(GET_POKEMON_TYPES);

    const tableData = useMemo(() => {
        const data = pokemonData?.pokemon_v2_pokemon || [];
        let result = data

        if (selectedTypes.length) {
            result = data.filter((pokemon: PokemonInterface) => {
                const pokemonTypes = pokemon.pokemon_v2_pokemontypes.map(({pokemon_v2_type}) => pokemon_v2_type.id);
                return pokemonTypes.some((typeId) => selectedTypes.includes(typeId))
            })
        }

        if (additionalFilters.useAdditionalFilters) {
            minMaxPokeFilters.forEach(filterName => {
                // @ts-ignore
                const filterValue = additionalFilters[filterName];

                if (filterValue.max <= 0) {
                    return null;
                }

                result = result.filter((pokemon: PokemonInterface) => {
                    // @ts-ignore
                    const pokemonValue = pokemon.pokemon_v2_pokemonspecy[filterName];
                    return inRange(pokemonValue, filterValue.min, filterValue.max + 1);
                })
            })
            boolPokeFilters.forEach(filterName => {
                // @ts-ignore
                const filterValue = additionalFilters[filterName];
                result = result.filter((pokemon: PokemonInterface) => {
                    // @ts-ignore
                    const pokemonValue = pokemon.pokemon_v2_pokemonspecy[filterName];
                    return pokemonValue === filterValue;
                })
            })
        }
        return result
    }, [pokemonData?.pokemon_v2_pokemon, selectedTypes, additionalFilters])

    return (
        <MainLayout>
            <HomePageStyles>
                <PokeStats pokemonTypes={pokemonTypes}/>
                <PokeFilter
                    setNameSearch={setNameSearch}
                    pokemonTypes={pokemonTypes}
                    setSelectedTypes={setSelectedTypes}
                    additionalFilters={additionalFilters}
                    setAdditionalFilters={setAdditionalFilters}
                />
                <LoaderWrapper loading={loading}>
                    <Table
                        className={'pika-table'}
                        columns={columns}
                        dataSource={tableData}
                        rowKey="name"
                    />
                </LoaderWrapper>
            </HomePageStyles>
        </MainLayout>
    )
};

export default HomePage;