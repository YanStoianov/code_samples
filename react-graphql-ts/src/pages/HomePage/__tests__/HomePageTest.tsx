import React from 'react'
import {render} from '@testing-library/react';
import 'jest-styled-components'
import HomePage from "../HomePage";
import {MockedProvider} from "@apollo/client/testing";
import {GET_POKEMON_TYPES} from "../../../queries/pokQueries";


const mocks = [
    {
        request: {
            query: GET_POKEMON_TYPES,
        },
        result: {
            data: {
                dog: {id: "1", name: "Buck", breed: "bulldog"}
            }
        }
    }
];

jest.mock('../../../../assets/pokedex-icon.jpg')

describe("HomePageTest", () => {
    beforeAll(() => {
        Object.defineProperty(window, "matchMedia", {
            writable: true,
            value: jest.fn().mockImplementation(query => ({
                matches: false,
                media: query,
                onchange: null,
                addListener: jest.fn(),
                removeListener: jest.fn(),
                addEventListener: jest.fn(),
                removeEventListener: jest.fn(),
                dispatchEvent: jest.fn(),
            }))
        });
    });

    test('snapshot', () => {
        const {asFragment} = render(
            <MockedProvider mocks={mocks} addTypename={false}>
                <HomePage/>
            </MockedProvider>
        );
        expect(asFragment()).toMatchSnapshot();
    })
});