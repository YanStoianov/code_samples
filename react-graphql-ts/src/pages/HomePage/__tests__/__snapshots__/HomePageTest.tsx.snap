// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HomePageTest snapshot 1`] = `
<DocumentFragment>
  <div
    class="sc-bcXHqe hFNikj"
  >
    <section
      class="ant-layout main-layout"
    >
      <header
        class="ant-layout-header layout-header"
      >
        <img
          alt="pokedex"
          class="layout-logo"
          src="[object Object]"
        />
        <h1
          class="layout-title"
        >
          Pokedex
        </h1>
      </header>
      <main
        class="ant-layout-content layout-content"
      >
        <div
          class="sc-dkrFOg fEXYpx"
        >
          <div
            class="sc-hLBbgP laINMI"
          >
            <div
              class="stats-block"
            >
              <div>
                Unique pokemons count: 0
              </div>
              <div>
                Pokemon duplications/variations count: 0
              </div>
              <div>
                Total pokemons count: 0
              </div>
              <div>
                Types count: 0
              </div>
            </div>
          </div>
          <div
            class="sc-eDvSVe kstY"
          >
            <div
              class="filters-block"
            >
              <input
                class="ant-input search"
                placeholder="Search"
                type="text"
                value=""
              />
              <div
                class="ant-select types-select ant-select-multiple ant-select-show-search"
              >
                <div
                  class="ant-select-selector"
                >
                  <div
                    class="ant-select-selection-overflow"
                  >
                    <div
                      class="ant-select-selection-overflow-item ant-select-selection-overflow-item-suffix"
                      style="opacity: 1;"
                    >
                      <div
                        class="ant-select-selection-search"
                        style="width: 0px;"
                      >
                        <input
                          aria-activedescendant="rc_select_TEST_OR_SSR_list_0"
                          aria-autocomplete="list"
                          aria-controls="rc_select_TEST_OR_SSR_list"
                          aria-haspopup="listbox"
                          aria-owns="rc_select_TEST_OR_SSR_list"
                          autocomplete="off"
                          class="ant-select-selection-search-input"
                          id="rc_select_TEST_OR_SSR"
                          readonly=""
                          role="combobox"
                          style="opacity: 0;"
                          type="search"
                          unselectable="on"
                          value=""
                        />
                        <span
                          aria-hidden="true"
                          class="ant-select-selection-search-mirror"
                        >
                          Â 
                        </span>
                      </div>
                    </div>
                  </div>
                  <span
                    class="ant-select-selection-placeholder"
                  >
                    Types
                  </span>
                </div>
              </div>
            </div>
            <div
              class="additional-cursor"
            >
              <div
                class="cursor-title"
              >
                Additional Filters
              </div>
              <span
                aria-label="down-square"
                class="anticon anticon-down-square"
                role="img"
                tabindex="-1"
              >
                <svg
                  aria-hidden="true"
                  data-icon="down-square"
                  fill="currentColor"
                  focusable="false"
                  height="1em"
                  viewBox="64 64 896 896"
                  width="1em"
                >
                  <path
                    d="M505.5 658.7c3.2 4.4 9.7 4.4 12.9 0l178-246c3.8-5.3 0-12.7-6.5-12.7H643c-10.2 0-19.9 4.9-25.9 13.2L512 558.6 406.8 413.2c-6-8.3-15.6-13.2-25.9-13.2H334c-6.5 0-10.3 7.4-6.5 12.7l178 246z"
                  />
                  <path
                    d="M880 112H144c-17.7 0-32 14.3-32 32v736c0 17.7 14.3 32 32 32h736c17.7 0 32-14.3 32-32V144c0-17.7-14.3-32-32-32zm-40 728H184V184h656v656z"
                  />
                </svg>
              </span>
            </div>
          </div>
          <div
            class="sc-gswNZR efVABA"
          >
            <div
              class="loading-block"
            >
              <div
                aria-busy="true"
                aria-live="polite"
                class="ant-spin ant-spin-spinning"
              >
                <span
                  aria-label="loading"
                  class="anticon anticon-loading anticon-spin ant-spin-dot"
                  role="img"
                  style="font-size: 4em;"
                >
                  <svg
                    aria-hidden="true"
                    data-icon="loading"
                    fill="currentColor"
                    focusable="false"
                    height="1em"
                    viewBox="0 0 1024 1024"
                    width="1em"
                  >
                    <path
                      d="M988 548c-19.9 0-36-16.1-36-36 0-59.4-11.6-117-34.6-171.3a440.45 440.45 0 00-94.3-139.9 437.71 437.71 0 00-139.9-94.3C629 83.6 571.4 72 512 72c-19.9 0-36-16.1-36-36s16.1-36 36-36c69.1 0 136.2 13.5 199.3 40.3C772.3 66 827 103 874 150c47 47 83.9 101.8 109.7 162.7 26.7 63.1 40.2 130.2 40.2 199.3.1 19.9-16 36-35.9 36z"
                    />
                  </svg>
                </span>
              </div>
            </div>
            <div
              class="child-blur"
            >
              <div
                class="ant-table-wrapper pika-table"
              >
                <div
                  class="ant-spin-nested-loading"
                >
                  <div
                    class="ant-spin-container"
                  >
                    <div
                      class="ant-table ant-table-empty"
                    >
                      <div
                        class="ant-table-container"
                      >
                        <div
                          class="ant-table-content"
                        >
                          <table
                            style="table-layout: auto;"
                          >
                            <colgroup>
                              <col
                                style="width: 200px;"
                              />
                              <col
                                style="width: 100px;"
                              />
                              <col
                                style="width: 120px;"
                              />
                              <col
                                style="width: 120px;"
                              />
                              <col
                                style="width: 120px;"
                              />
                              <col
                                style="width: 120px;"
                              />
                              <col
                                style="width: 120px;"
                              />
                              <col
                                style="width: 120px;"
                              />
                              <col
                                style="width: 450px;"
                              />
                            </colgroup>
                            <thead
                              class="ant-table-thead"
                            >
                              <tr>
                                <th
                                  aria-sort="ascending"
                                  class="ant-table-cell ant-table-column-sort ant-table-column-has-sorters"
                                  tabindex="0"
                                >
                                  <div
                                    class="ant-table-column-sorters"
                                  >
                                    <span
                                      class="ant-table-column-title"
                                    >
                                      Name
                                    </span>
                                    <span
                                      class="ant-table-column-sorter ant-table-column-sorter-full"
                                    >
                                      <span
                                        class="ant-table-column-sorter-inner"
                                      >
                                        <span
                                          aria-label="caret-up"
                                          class="anticon anticon-caret-up ant-table-column-sorter-up active"
                                          role="presentation"
                                        >
                                          <svg
                                            aria-hidden="true"
                                            data-icon="caret-up"
                                            fill="currentColor"
                                            focusable="false"
                                            height="1em"
                                            viewBox="0 0 1024 1024"
                                            width="1em"
                                          >
                                            <path
                                              d="M858.9 689L530.5 308.2c-9.4-10.9-27.5-10.9-37 0L165.1 689c-12.2 14.2-1.2 35 18.5 35h656.8c19.7 0 30.7-20.8 18.5-35z"
                                            />
                                          </svg>
                                        </span>
                                        <span
                                          aria-label="caret-down"
                                          class="anticon anticon-caret-down ant-table-column-sorter-down"
                                          role="presentation"
                                        >
                                          <svg
                                            aria-hidden="true"
                                            data-icon="caret-down"
                                            fill="currentColor"
                                            focusable="false"
                                            height="1em"
                                            viewBox="0 0 1024 1024"
                                            width="1em"
                                          >
                                            <path
                                              d="M840.4 300H183.6c-19.7 0-30.7 20.8-18.5 35l328.4 380.8c9.4 10.9 27.5 10.9 37 0L858.9 335c12.2-14.2 1.2-35-18.5-35z"
                                            />
                                          </svg>
                                        </span>
                                      </span>
                                    </span>
                                  </div>
                                </th>
                                <th
                                  class="ant-table-cell"
                                >
                                  Types
                                </th>
                                <th
                                  aria-label="hp sortable"
                                  class="ant-table-cell ant-table-column-has-sorters"
                                  tabindex="0"
                                >
                                  <div
                                    class="ant-table-column-sorters"
                                  >
                                    <span
                                      class="ant-table-column-title"
                                    >
                                      hp
                                    </span>
                                    <span
                                      class="ant-table-column-sorter ant-table-column-sorter-full"
                                    >
                                      <span
                                        class="ant-table-column-sorter-inner"
                                      >
                                        <span
                                          aria-label="caret-up"
                                          class="anticon anticon-caret-up ant-table-column-sorter-up"
                                          role="presentation"
                                        >
                                          <svg
                                            aria-hidden="true"
                                            data-icon="caret-up"
                                            fill="currentColor"
                                            focusable="false"
                                            height="1em"
                                            viewBox="0 0 1024 1024"
                                            width="1em"
                                          >
                                            <path
                                              d="M858.9 689L530.5 308.2c-9.4-10.9-27.5-10.9-37 0L165.1 689c-12.2 14.2-1.2 35 18.5 35h656.8c19.7 0 30.7-20.8 18.5-35z"
                                            />
                                          </svg>
                                        </span>
                                        <span
                                          aria-label="caret-down"
                                          class="anticon anticon-caret-down ant-table-column-sorter-down"
                                          role="presentation"
                                        >
                                          <svg
                                            aria-hidden="true"
                                            data-icon="caret-down"
                                            fill="currentColor"
                                            focusable="false"
                                            height="1em"
                                            viewBox="0 0 1024 1024"
                                            width="1em"
                                          >
                                            <path
                                              d="M840.4 300H183.6c-19.7 0-30.7 20.8-18.5 35l328.4 380.8c9.4 10.9 27.5 10.9 37 0L858.9 335c12.2-14.2 1.2-35-18.5-35z"
                                            />
                                          </svg>
                                        </span>
                                      </span>
                                    </span>
                                  </div>
                                </th>
                                <th
                                  aria-label="attack sortable"
                                  class="ant-table-cell ant-table-column-has-sorters"
                                  tabindex="0"
                                >
                                  <div
                                    class="ant-table-column-sorters"
                                  >
                                    <span
                                      class="ant-table-column-title"
                                    >
                                      attack
                                    </span>
                                    <span
                                      class="ant-table-column-sorter ant-table-column-sorter-full"
                                    >
                                      <span
                                        class="ant-table-column-sorter-inner"
                                      >
                                        <span
                                          aria-label="caret-up"
                                          class="anticon anticon-caret-up ant-table-column-sorter-up"
                                          role="presentation"
                                        >
                                          <svg
                                            aria-hidden="true"
                                            data-icon="caret-up"
                                            fill="currentColor"
                                            focusable="false"
                                            height="1em"
                                            viewBox="0 0 1024 1024"
                                            width="1em"
                                          >
                                            <path
                                              d="M858.9 689L530.5 308.2c-9.4-10.9-27.5-10.9-37 0L165.1 689c-12.2 14.2-1.2 35 18.5 35h656.8c19.7 0 30.7-20.8 18.5-35z"
                                            />
                                          </svg>
                                        </span>
                                        <span
                                          aria-label="caret-down"
                                          class="anticon anticon-caret-down ant-table-column-sorter-down"
                                          role="presentation"
                                        >
                                          <svg
                                            aria-hidden="true"
                                            data-icon="caret-down"
                                            fill="currentColor"
                                            focusable="false"
                                            height="1em"
                                            viewBox="0 0 1024 1024"
                                            width="1em"
                                          >
                                            <path
                                              d="M840.4 300H183.6c-19.7 0-30.7 20.8-18.5 35l328.4 380.8c9.4 10.9 27.5 10.9 37 0L858.9 335c12.2-14.2 1.2-35-18.5-35z"
                                            />
                                          </svg>
                                        </span>
                                      </span>
                                    </span>
                                  </div>
                                </th>
                                <th
                                  aria-label="defense sortable"
                                  class="ant-table-cell ant-table-column-has-sorters"
                                  tabindex="0"
                                >
                                  <div
                                    class="ant-table-column-sorters"
                                  >
                                    <span
                                      class="ant-table-column-title"
                                    >
                                      defense
                                    </span>
                                    <span
                                      class="ant-table-column-sorter ant-table-column-sorter-full"
                                    >
                                      <span
                                        class="ant-table-column-sorter-inner"
                                      >
                                        <span
                                          aria-label="caret-up"
                                          class="anticon anticon-caret-up ant-table-column-sorter-up"
                                          role="presentation"
                                        >
                                          <svg
                                            aria-hidden="true"
                                            data-icon="caret-up"
                                            fill="currentColor"
                                            focusable="false"
                                            height="1em"
                                            viewBox="0 0 1024 1024"
                                            width="1em"
                                          >
                                            <path
                                              d="M858.9 689L530.5 308.2c-9.4-10.9-27.5-10.9-37 0L165.1 689c-12.2 14.2-1.2 35 18.5 35h656.8c19.7 0 30.7-20.8 18.5-35z"
                                            />
                                          </svg>
                                        </span>
                                        <span
                                          aria-label="caret-down"
                                          class="anticon anticon-caret-down ant-table-column-sorter-down"
                                          role="presentation"
                                        >
                                          <svg
                                            aria-hidden="true"
                                            data-icon="caret-down"
                                            fill="currentColor"
                                            focusable="false"
                                            height="1em"
                                            viewBox="0 0 1024 1024"
                                            width="1em"
                                          >
                                            <path
                                              d="M840.4 300H183.6c-19.7 0-30.7 20.8-18.5 35l328.4 380.8c9.4 10.9 27.5 10.9 37 0L858.9 335c12.2-14.2 1.2-35-18.5-35z"
                                            />
                                          </svg>
                                        </span>
                                      </span>
                                    </span>
                                  </div>
                                </th>
                                <th
                                  aria-label="special-attack sortable"
                                  class="ant-table-cell ant-table-column-has-sorters"
                                  tabindex="0"
                                >
                                  <div
                                    class="ant-table-column-sorters"
                                  >
                                    <span
                                      class="ant-table-column-title"
                                    >
                                      special-attack
                                    </span>
                                    <span
                                      class="ant-table-column-sorter ant-table-column-sorter-full"
                                    >
                                      <span
                                        class="ant-table-column-sorter-inner"
                                      >
                                        <span
                                          aria-label="caret-up"
                                          class="anticon anticon-caret-up ant-table-column-sorter-up"
                                          role="presentation"
                                        >
                                          <svg
                                            aria-hidden="true"
                                            data-icon="caret-up"
                                            fill="currentColor"
                                            focusable="false"
                                            height="1em"
                                            viewBox="0 0 1024 1024"
                                            width="1em"
                                          >
                                            <path
                                              d="M858.9 689L530.5 308.2c-9.4-10.9-27.5-10.9-37 0L165.1 689c-12.2 14.2-1.2 35 18.5 35h656.8c19.7 0 30.7-20.8 18.5-35z"
                                            />
                                          </svg>
                                        </span>
                                        <span
                                          aria-label="caret-down"
                                          class="anticon anticon-caret-down ant-table-column-sorter-down"
                                          role="presentation"
                                        >
                                          <svg
                                            aria-hidden="true"
                                            data-icon="caret-down"
                                            fill="currentColor"
                                            focusable="false"
                                            height="1em"
                                            viewBox="0 0 1024 1024"
                                            width="1em"
                                          >
                                            <path
                                              d="M840.4 300H183.6c-19.7 0-30.7 20.8-18.5 35l328.4 380.8c9.4 10.9 27.5 10.9 37 0L858.9 335c12.2-14.2 1.2-35-18.5-35z"
                                            />
                                          </svg>
                                        </span>
                                      </span>
                                    </span>
                                  </div>
                                </th>
                                <th
                                  aria-label="special-defense sortable"
                                  class="ant-table-cell ant-table-column-has-sorters"
                                  tabindex="0"
                                >
                                  <div
                                    class="ant-table-column-sorters"
                                  >
                                    <span
                                      class="ant-table-column-title"
                                    >
                                      special-defense
                                    </span>
                                    <span
                                      class="ant-table-column-sorter ant-table-column-sorter-full"
                                    >
                                      <span
                                        class="ant-table-column-sorter-inner"
                                      >
                                        <span
                                          aria-label="caret-up"
                                          class="anticon anticon-caret-up ant-table-column-sorter-up"
                                          role="presentation"
                                        >
                                          <svg
                                            aria-hidden="true"
                                            data-icon="caret-up"
                                            fill="currentColor"
                                            focusable="false"
                                            height="1em"
                                            viewBox="0 0 1024 1024"
                                            width="1em"
                                          >
                                            <path
                                              d="M858.9 689L530.5 308.2c-9.4-10.9-27.5-10.9-37 0L165.1 689c-12.2 14.2-1.2 35 18.5 35h656.8c19.7 0 30.7-20.8 18.5-35z"
                                            />
                                          </svg>
                                        </span>
                                        <span
                                          aria-label="caret-down"
                                          class="anticon anticon-caret-down ant-table-column-sorter-down"
                                          role="presentation"
                                        >
                                          <svg
                                            aria-hidden="true"
                                            data-icon="caret-down"
                                            fill="currentColor"
                                            focusable="false"
                                            height="1em"
                                            viewBox="0 0 1024 1024"
                                            width="1em"
                                          >
                                            <path
                                              d="M840.4 300H183.6c-19.7 0-30.7 20.8-18.5 35l328.4 380.8c9.4 10.9 27.5 10.9 37 0L858.9 335c12.2-14.2 1.2-35-18.5-35z"
                                            />
                                          </svg>
                                        </span>
                                      </span>
                                    </span>
                                  </div>
                                </th>
                                <th
                                  aria-label="speed sortable"
                                  class="ant-table-cell ant-table-column-has-sorters"
                                  tabindex="0"
                                >
                                  <div
                                    class="ant-table-column-sorters"
                                  >
                                    <span
                                      class="ant-table-column-title"
                                    >
                                      speed
                                    </span>
                                    <span
                                      class="ant-table-column-sorter ant-table-column-sorter-full"
                                    >
                                      <span
                                        class="ant-table-column-sorter-inner"
                                      >
                                        <span
                                          aria-label="caret-up"
                                          class="anticon anticon-caret-up ant-table-column-sorter-up"
                                          role="presentation"
                                        >
                                          <svg
                                            aria-hidden="true"
                                            data-icon="caret-up"
                                            fill="currentColor"
                                            focusable="false"
                                            height="1em"
                                            viewBox="0 0 1024 1024"
                                            width="1em"
                                          >
                                            <path
                                              d="M858.9 689L530.5 308.2c-9.4-10.9-27.5-10.9-37 0L165.1 689c-12.2 14.2-1.2 35 18.5 35h656.8c19.7 0 30.7-20.8 18.5-35z"
                                            />
                                          </svg>
                                        </span>
                                        <span
                                          aria-label="caret-down"
                                          class="anticon anticon-caret-down ant-table-column-sorter-down"
                                          role="presentation"
                                        >
                                          <svg
                                            aria-hidden="true"
                                            data-icon="caret-down"
                                            fill="currentColor"
                                            focusable="false"
                                            height="1em"
                                            viewBox="0 0 1024 1024"
                                            width="1em"
                                          >
                                            <path
                                              d="M840.4 300H183.6c-19.7 0-30.7 20.8-18.5 35l328.4 380.8c9.4 10.9 27.5 10.9 37 0L858.9 335c12.2-14.2 1.2-35-18.5-35z"
                                            />
                                          </svg>
                                        </span>
                                      </span>
                                    </span>
                                  </div>
                                </th>
                                <th
                                  class="ant-table-cell"
                                >
                                  Species
                                </th>
                              </tr>
                            </thead>
                            <tbody
                              class="ant-table-tbody"
                            >
                              <tr
                                class="ant-table-placeholder"
                              >
                                <td
                                  class="ant-table-cell"
                                  colspan="9"
                                >
                                  <div
                                    class="ant-empty ant-empty-normal"
                                  >
                                    <div
                                      class="ant-empty-image"
                                    >
                                      <svg
                                        class="ant-empty-img-simple"
                                        height="41"
                                        viewBox="0 0 64 41"
                                        width="64"
                                        xmlns="http://www.w3.org/2000/svg"
                                      >
                                        <g
                                          fill="none"
                                          fill-rule="evenodd"
                                          transform="translate(0 1)"
                                        >
                                          <ellipse
                                            class="ant-empty-img-simple-ellipse"
                                            cx="32"
                                            cy="33"
                                            rx="32"
                                            ry="7"
                                          />
                                          <g
                                            class="ant-empty-img-simple-g"
                                            fill-rule="nonzero"
                                          >
                                            <path
                                              d="M55 12.76L44.854 1.258C44.367.474 43.656 0 42.907 0H21.093c-.749 0-1.46.474-1.947 1.257L9 12.761V22h46v-9.24z"
                                            />
                                            <path
                                              class="ant-empty-img-simple-path"
                                              d="M41.613 15.931c0-1.605.994-2.93 2.227-2.931H55v18.137C55 33.26 53.68 35 52.05 35h-40.1C10.32 35 9 33.259 9 31.137V13h11.16c1.233 0 2.227 1.323 2.227 2.928v.022c0 1.605 1.005 2.901 2.237 2.901h14.752c1.232 0 2.237-1.308 2.237-2.913v-.007z"
                                            />
                                          </g>
                                        </g>
                                      </svg>
                                    </div>
                                    <div
                                      class="ant-empty-description"
                                    >
                                      No Data
                                    </div>
                                  </div>
                                </td>
                              </tr>
                            </tbody>
                          </table>
                        </div>
                      </div>
                    </div>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
      </main>
      <footer
        class="ant-layout-footer layout-footer"
      >
        Â©2022 Created by Vlad
      </footer>
    </section>
  </div>
</DocumentFragment>
`;
